src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8Utility.cs

```diff
@@ -44,6 +60,143 @@ public static unsafe int GetIndexOfFirstInvalidUtf8Sequence(ReadOnlySpan<byte> u
         }
 
 #if FEATURE_UTF8STRING
+        /// <summary>
+        /// Given a buffer <paramref name="utf8Data"/> which represents guaranteed well-formed UTF-8 data,
+        /// returns the number of <see langword="char"/>s which would result from transcoding the buffer
+        /// to UTF-16. The behavior of this method is undefined if <paramref name="utf8Data"/> contains
+        /// any ill-formed UTF-8 subsequences.
+        /// </summary>
+        public static unsafe int GetUtf16CharCountFromKnownWellFormedUtf8(ReadOnlySpan<byte> utf8Data)
+        {
+            // Remember: the number of resulting UTF-16 chars will never be greater than the number
+            // of UTF-8 bytes given well-formed input, so we can get away with casting the final
+            // result to an 'int'.
+
+            fixed (byte* pPinnedUtf8Data = &MemoryMarshal.GetReference(utf8Data))
+            {
+                if (Sse2.IsSupported && Popcnt.IsSupported)
+                {
+                    // Optimizations via SSE2 & POPCNT are available - use them.
+
+                    Debug.Assert(BitConverter.IsLittleEndian, "SSE2 only supported on little-endian platforms.");
+                    Debug.Assert(sizeof(nint) == IntPtr.Size, "nint defined incorrectly.");
+                    Debug.Assert(sizeof(nuint) == IntPtr.Size, "nuint defined incorrectly.");
+
+                    byte* pBuffer = pPinnedUtf8Data;
+                    nuint bufferLength = (uint)utf8Data.Length;
+
+                    // Optimization: Can we stay in the all-ASCII code paths?
+
+                    nuint utf16CharCount = ASCIIUtility.GetIndexOfFirstNonAsciiByte(pBuffer, bufferLength);
+
+                    if (utf16CharCount != bufferLength)
+                    {
+                        // Found at least one non-ASCII byte, so fall down the slower (but still vectorized) code paths.
+                        // Given well-formed UTF-8 input, we can compute the number of resulting UTF-16 code units
+                        // using the following formula:
+                        //
+                        // utf16CharCount = utf8ByteCount - numUtf8ContinuationBytes + numUtf8FourByteHeaders
+
+                        utf16CharCount = bufferLength;
+
+                        Vector128<sbyte> vecAllC0 = Vector128.Create(unchecked((sbyte)0xC0));
+                        Vector128<sbyte> vecAll80 = Vector128.Create(unchecked((sbyte)0x80));
+                        Vector128<sbyte> vecAll6F = Vector128.Create(unchecked((sbyte)0x6F));
```

Comment:
>unchecked [](start = 69, length = 9)

you don't need the uncheck here I guess

Reply:
```
Constant value '192' cannot be converted to a 'sbyte' (use 'unchecked' syntax to override)
```
It's needed.

Reply:
0x6F is 111 and not 192

---
In reply to: [332667138](https://github.com/dotnet/coreclr/pull/27078#discussion_r332667138) [](ancestors = 332667138)

Reply:
Ah, you meant just the last one (I checked the first (192)). You are right, sorry.